### PR TITLE
Fix crash during server render in react 16.4.1.

### PR DESCRIPTION
**Pull request by dustinsoftware**
This fixes a crash that only occurs in react-dom 16.4.1.

setTimeout and clearTimeout may not be available in some server-render environments (such as ChakraCore in React.NET), and loading ReactScheduler.js will cause a crash unless the existence of the variables are checked via a typeof comparison.

https://github.com/reactjs/React.NET/issues/555

The crash did not occur in 16.4.0, and the change appears to have been introduced here: https://github.com/facebook/react/pull/12931/files#diff-bbebc3357e1fb99ab13ad796e04b69a6L47

I tested this by using yarn link and running it with a local copy of React.NET. I am unsure the best way to unit test this change, since assigning null to `setTimeout` causes an immediate crash within the Node REPL.

### CODE REVIEW

**Comment by gaearon** [comment]
Is there any way you can shim it on your side? We could take this but then it's highly likely we'll break it again. It's also plausible we'll start relying on e.g. Promise so if your environment is very unusual then you'll keep running into those issues.

**Comment by dustinsoftware** [comment]
Thanks for the fast response. I'll investigate adding a shim on React.NET for those functions and expect that it will break again in the future. Would you be open to accepting the patch anyway so we stop getting crashes in the meantime?

I opened a PR directly since this one took a similar approach: https://github.com/facebook/react/issues/13000

I'll look at fixing the flow errors, do you have any suggestions on what to do? Should we return a function that warns instead?

cc @Daniel15

**Comment by Daniel15** [comment]
> Is there any way you can shim it on your side? 

@gaearon - Relying on `setTimeout` and `setInterval` is going to break any non-browser rendering environment where the DOM/BOM is not available. Is it expected that `ReactScheduler` is only loaded in browser environments?

**Comment by Daniel15** [comment]
>  if your environment is very unusual then you'll keep running into those issues.

ReactJS.NET uses a pretty standard V8 or ChakraCore environment.

**Comment by Daniel15** [comment]
@dustinsoftware  In the meantime we can patch the version of React bundled with ReactJS.NET (although that means it'll still be broken for people that load their own version of React rather than using the bundled one). Either that or just shim `setTimeout` and `setInterval`, but that seems very hacky and may break other stuff. Since React seems to be pulling them into a variable, maybe shim it when React is being loaded, then unshim it once React has loaded?

**Comment by react-linter** [comment]
<details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 5b3d17a5f729eb204371ad1953a07ec3ae3dae00...d29667acc5c9cf45aef228cbdd9ecf2d01ebc784</p>


  
## react-scheduler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-scheduler.development.js | +1.8% | +0.9% | 15.1 KB | 15.38 KB | 4.73 KB | 4.78 KB | UMD_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.1%** | **:small_red_triangle:+1.3%** | **2.43 KB** | **2.51 KB** | **1.18 KB** | **1.2 KB** | **UMD_PROD**
react-scheduler.development.js | +1.8% | +0.9% | 14.91 KB | 15.19 KB | 4.69 KB | 4.73 KB | NODE_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.3%** | **:small_red_triangle:+1.4%** | **2.34 KB** | **2.42 KB** | **1.13 KB** | **1.14 KB** | **NODE_PROD**
ReactScheduler-dev.js | +2.5% | +1.3% | 13.67 KB | 14.01 KB | 4.21 KB | 4.27 KB | FB_WWW_DEV
**ReactScheduler-prod.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+0.8%** | **6.91 KB** | **7 KB** | **1.8 KB** | **1.82 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by gaearon** [comment]
>Relying on setTimeout and setInterval is going to break any non-browser rendering environment where the DOM/BOM is not available

We generally target Node environment with the server renderer. Node also has setTimeout.

I don’t think the server side renderer uses the scheduler. You’re likely seeing this issue because some component imports regular `react-dom` (e.g. for `findDOMNode`) which ends up pulling this initialization code in.

File: packages/react-scheduler/src/ReactScheduler.js [hunk 1]
```diff
@@ -60,8 +60,21 @@ if (__DEV__) {
 // this module is initially evaluated.
 // We want to be using a consistent implementation.
 const localDate = Date;
-const localSetTimeout = setTimeout;
-const localClearTimeout = clearTimeout;
+const localSetTimeout =
+  typeof setTimeout === 'function'
+    ? setTimeout
+    : warning(
```
Comment (gaearon) [inline packages/react-scheduler/src/ReactScheduler.js]: This looks a bit misleading, as if we’re using its return value.

Let’s just keep undefined there explicitly and then warn later.

File: packages/react-scheduler/src/ReactScheduler.js [hunk 2]
```diff
@@ -104,6 +94,11 @@ if (!canUseDOM) {
     callback: FrameCallbackType,
     options?: {timeout: number},
   ): CallbackIdType {
+    invariant(
+      typeof localSetTimeout === 'function',
+      'localSetTimeout is not a function. Please load a polyfill that defines this function.',
```
Comment (gaearon) [inline packages/react-scheduler/src/ReactScheduler.js]: `localSetTimeout` is a variable name so it doesn't mean anything to somebody reading this. That said, I don't think we want an invariant at all.

**Comment by dustinsoftware** [comment]
Ah, yes I did add that because of flow. I’ll push a fix.

**Comment by dustinsoftware** [comment]
I may be misunderstanding but it sounds like you prefer to have error
message at all, so that’s what I pushed. If I misunderstood I can restore
the warning inside an ”if dev mode” block so it is optimized out in
production mode.

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.1%, gzip: :small_red_triangle:+0.1%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 5b3d17a5f729eb204371ad1953a07ec3ae3dae00...96881ff600f59f3945b757eeaa2afe7993f08997</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.1% | 626.73 KB | 627.06 KB | 146.5 KB | 146.62 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **95.3 KB** | **95.38 KB** | **30.82 KB** | **30.84 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | +0.1% | 622.86 KB | 623.19 KB | 145.33 KB | 145.45 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **95.29 KB** | **95.37 KB** | **30.35 KB** | **30.36 KB** | **NODE_PROD**
react-dom.profiling.min.js | +0.1% | 0.0% | 96.26 KB | 96.34 KB | 30.68 KB | 30.69 KB | NODE_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.1% | +0.1% | 410.78 KB | 411.11 KB | 92.08 KB | 92.19 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **82.49 KB** | **82.57 KB** | **25.42 KB** | **25.42 KB** | **UMD_PROD**
react-art.development.js | +0.1% | +0.2% | 342.76 KB | 343.09 KB | 74.94 KB | 75.06 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.1%** | **47.47 KB** | **47.55 KB** | **14.79 KB** | **14.8 KB** | **NODE_PROD**

## react-scheduler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-scheduler.development.js | +2.2% | +2.4% | 15.1 KB | 15.43 KB | 4.73 KB | 4.85 KB | UMD_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.1%** | **:small_red_triangle:+1.3%** | **2.43 KB** | **2.51 KB** | **1.18 KB** | **1.2 KB** | **UMD_PROD**
react-scheduler.development.js | +2.2% | +2.4% | 14.91 KB | 15.24 KB | 4.69 KB | 4.8 KB | NODE_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.3%** | **:small_red_triangle:+1.4%** | **2.34 KB** | **2.42 KB** | **1.13 KB** | **1.14 KB** | **NODE_PROD**
ReactScheduler-dev.js | +2.4% | +2.7% | 13.67 KB | 14 KB | 4.21 KB | 4.33 KB | FB_WWW_DEV
**ReactScheduler-prod.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+0.8%** | **6.91 KB** | **7 KB** | **1.8 KB** | **1.82 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request merged**